File: sdk/elixir/lib/rocksky/shout.ex
defmodule Rocksky.Shout do
  @moduledoc "`app.rocksky.shout.*` endpoints."

  alias Rocksky.HTTP

  @doc "Create a shout. Body: `:message`."
  def create_shout(client, body),
    do: HTTP.procedure(client, "app.rocksky.shout.createShout", [], Map.new(body))

  @doc "Reply to a shout. Body: `:shoutId`, `:message`."
  def reply_shout(client, body),
    do: HTTP.procedure(client, "app.rocksky.shout.replyShout", [], Map.new(body))

  @doc "Remove a shout. Params: `:id`."
  def remove_shout(client, params),
    do: HTTP.procedure(client, "app.rocksky.shout.removeShout", params)

  @doc "Report a shout. Body: `:shoutId`, `:reason`."
  def report_shout(client, body),
    do: HTTP.procedure(client, "app.rocksky.shout.reportShout", [], Map.new(body))

  @doc "Shouts on a profile. Params: `:did`, `:limit`, `:offset`."
  def get_profile_shouts(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.shout.getProfileShouts", params)

  @doc "Shouts on an album. Params: `:uri`, `:limit`, `:offset`."
  def get_album_shouts(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.shout.getAlbumShouts", params)

  @doc "Shouts on an artist. Params: `:uri`, `:limit`, `:offset`."
  def get_artist_shouts(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.shout.getArtistShouts", params)

  @doc "Shouts on a track. Params: `:uri`."
  def get_track_shouts(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.shout.getTrackShouts", params)

  @doc "Replies to a shout. Params: `:uri`, `:limit`, `:offset`."
  def get_shout_replies(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.shout.getShoutReplies", params)
end

File: sdk/clojure/src/rocksky/shout.clj
(ns rocksky.shout
  "Endpoints under app.rocksky.shout.*"
  (:require [rocksky.client :as c]))

(defn create-shout
  "Create a new shout. Required: `:message`."
  [client {:keys [message]}]
  (c/procedure client :app.rocksky.shout.createShout {:message message}))

(defn reply-shout
  "Reply to a shout. Required: `:shout-id` `:message`."
  [client {:keys [shout-id message]}]
  (c/procedure client :app.rocksky.shout.replyShout
               {:shoutId shout-id :message message}))

(defn remove-shout
  "Remove a shout. Required: `:id`."
  [client {:keys [id]}]
  (c/procedure client :app.rocksky.shout.removeShout nil {:id id}))

(defn report-shout
  "Report a shout. Required: `:shout-id`. Optional: `:reason`."
  [client {:keys [shout-id reason]}]
  (c/procedure client :app.rocksky.shout.reportShout
               (cond-> {:shoutId shout-id}
                 reason (assoc :reason reason))))

(defn get-album-shouts
  "Get shouts for an album.

  Required: `:uri`. Optional: `:limit` `:offset`."
  [client {:keys [uri limit offset]}]
  (c/query client :app.rocksky.shout.getAlbumShouts
           {:uri uri :limit limit :offset offset}))

(defn get-artist-shouts
  "Get shouts for an artist.

  Required: `:uri`. Optional: `:limit` `:offset`."
  [client {:keys [uri limit offset]}]
  (c/query client :app.rocksky.shout.getArtistShouts
           {:uri uri :limit limit :offset offset}))

(defn get-track-shouts
  "Get all shouts for a specific track. Required: `:uri`."
  [client {:keys [uri]}]
  (c/query client :app.rocksky.shout.getTrackShouts {:uri uri}))

(defn get-profile-shouts
  "Get shouts on an actor's profile.

  Required: `:did`. Optional: `:limit` `:offset`."
  [client {:keys [did limit offset]}]
  (c/query client :app.rocksky.shout.getProfileShouts
           {:did did :limit limit :offset offset}))

(defn get-shout-replies
  "Get replies to a shout.

  Required: `:uri`. Optional: `:limit` `:offset`."
  [client {:keys [uri limit offset]}]
  (c/query client :app.rocksky.shout.getShoutReplies
           {:uri uri :limit limit :offset offset}))
